<a href="https://colab.research.google.com/github/Gabriel-PC/Desafio-DATA_Viking---Detector-de-Fumaca/blob/main/Desafio_DIO_An%C3%A1lise_de_dados_com_Python_e_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
#import pandas e matplotlib
import pandas as pd
import matplotlib.pyplot as plt
import warnings
import statistics
from pandas._libs.hashtable import value_count
from statistics import mean 


In [4]:
#importando a base de dados
data1AW = pd.read_excel("/content/AdventureWorks.xlsx")
data2GM = pd.read_csv("/content/Gapminder.csv", error_bad_lines=False, sep=";")

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [14]:
data2GM = data2GM.rename(columns={"country":"Pais", "continent":"Continente", "year":"Ano", "lifeExp":"ExpecVida", "pop":"Populacao", "gdpPercap":"PIB per Capita" })
data2GM

Pais Continente   Ano            ExpecVida  Populacao  \
0         Greece     Europe  1950                   66    7566028   
1       Portugal     Europe  1950                   59    8442750   
2      Sri Lanka       Asia  1950                   56    7533097   
3         Russia        FSU  1950                   57  101936816   
4         Belize        NaN  1950                   58      65797   
...          ...        ...   ...                  ...        ...   
3307  Mauritania     Africa  2007                   64    3270065   
3308   Mauritius     Africa  2007                   73    1250882   
3309      Mexico   Americas  2007                   76  108700891   
3310        Mali     Africa  2007                   54   12031795   
3311    Zimbabwe     Africa  2007                   43   12311143   

           PIB per Capita                  PIB  
0                   3,293       24,915,193,297  
1                   2,962       25,006,502,336  
2                   1,033        7,785,448,262  
3                   3,675      374,612,976,271  
4                   1,731          113,903,340  
...                   ...                  ...  
3307                1,803        5,896,422,597  
3308               10,957       13,705,902,966  
3309               11,978    1,301,973,070,171  
3310                1,043       12,544,127,565  
3311                  470        5,782,658,337  

[3312 rows x 7 columns]

In [ ]:
# Verificando campos nulos
print(data2GM.isna().sum())
# Verificando a consistência dos dados (max, min, count): temperatura, umidade relativa, pressão, alarme: ok
print(data2GM.describe(include = 'all'))
# Verificando dados únicos: alarme, tempo e 'index': ok
print(data2GM.nunique().sort_values())
# Removendo as linhas com valores duplicados
data2GM = data2GM.drop_duplicates()
print(data2GM) #esperado já que os tempos são todos únicos

In [19]:
#Acrescentando uma coluna calculada a planilha
data2GM["PIB"] = data2GM["PIB per Capita"].mul(data2GM["Populacao"])

#Ajustando a apresentação dos dados de float muito grandes
pd.options.display.float_format="{:20,.0f}".format
pd.set_option('display.max_columns', 30)
warnings.filterwarnings('ignore')
#data2_GM.head()
data2GM["Ano"] = pd.to_datetime(data2GM["Ano"], format='%Y')
data2GM['Ano'] = pd.DatetimeIndex(data2GM["Ano"]).year
#data2_GM.dtypes
data2GM.describe()

Ano            ExpecVida            Populacao  \
count                3,312                3,312                3,312   
mean                 1,980                   65           31,614,891   
std                     17                   12          104,119,343   
min                  1,950                   24               59,412   
25%                  1,967                   58            2,678,572   
50%                  1,982                   70            7,557,218   
75%                  1,996                   74           19,585,222   
max                  2,007                   83        1,318,683,096   

            PIB per Capita                  PIB  
count                3,312                3,312  
mean                11,317      304,922,654,087  
std                 11,369      981,982,327,978  
min                    241           52,784,691  
25%                  2,515        8,988,503,786  
50%                  7,839       49,817,768,152  
75%                 17,358      185,359,011,956  
max                113,523   12,934,458,535,085

In [32]:
data2GM['Ano'].describe()

count                  3,312
mean                   1,980
std                       17
min                    1,950
25%                    1,967
50%                    1,982
75%                    1,996
max                    2,007
Name: Ano, dtype: float64

In [47]:
data2GM.mean()

Ano                             1,980
ExpecVida                          65
Populacao                  31,614,891
PIB per Capita                 11,317
PIB                   304,922,654,087
dtype: float64

In [48]:
data2GM["PIB"].mean()

304922654086.72473

In [44]:
data2GM.groupby(['Ano'])['PIB'].mean()

435285011494.7466

In [ ]:
data2GM = data2GM.sort_values("Ano", ascending=True).reset_index(drop=True)
data2GM

In [46]:
data2GM.groupby(['Ano']).mean()['PIB per Capita'].mean()

14386.002855136041

In [34]:
statistics.mean(data2GM.groupby(['Ano']).mean()['PIB'])

435285011494.7465

In [ ]:
from pandas._libs.hashtable import value_count
#Apresentando os 20% mais ricos e os 20% mais pobres
import numpy as np

#print(np.mean(data2_GM.sort_values(by=["Ano","PIB per Capita"]).reset_index(drop=True).head(662)))
#data2_GM = data2_GM.sort_values(by=["Ano","PIB per Capita"]).reset_index(drop=True)
#data2_GM.groupby(by = ['Ano','Continente', 'Pais'])['PIB per Capita'].mean()
#for Ano in (data2_GM['Ano']).unique() : print(Ano, np.mean(data2_GM["PIB per Capita"]))
#print(np.mean(pd.DataFrame(data2_GM["Ano"])).sort_values(by=["Ano","PIB per Capita"]).reset_index(drop=True).tail(662))

contador_por_ano = value_count(data2GM["Ano"] == "1950")

for x in contador_por_ano :
  print(np.mean(data2GM.at[x, 5]))
# quero o elemento que está: 
# na 3ª linha (índice 2)
# na 1ª coluna (índice 0)
#print(df.iat[2, 0])


In [ ]:
print(data2GM[data2GM["Pais"] == "Libya"])
print(np.mean(data2GM[data2GM["Pais"] == "Libya"]))

In [ ]:
#for Ano in (data2_GM["Ano"]) : print(np.mean(data2_GM[data2_GM["PIB per Capita"] == Ano]))

for Ano in data2GM["Ano"].unique() : print(data2GM.groupby(by = ['Ano','Continente', 'Pais'])['PIB per Capita'].mean())

#data2_GM.groupby(by = ['Ano','Continente', 'Pais'])['PIB per Capita'].mean()

In [ ]:
magicians = ['alice', 'david', 'carolina']
for magician in magicians: print(magician)

In [43]:
print(data2GM["Ano"].unique())
#print((data2_GM.groupby(by = ['Ano', 'Continente', 'Pais'])['PIB per Capita']).count())
print(data2GM.query("Ano").mean())
#df.query('(NU_ANO == 2015) & (SG_UF_ESCOLA == "BA") & (NU_MEDIA_MT >= 600)')
#df.loc[df['NU_ANO'] == 2015]
#df.loc[df['NU_ANO'] == 2015]['TP_DEPENDENCIA_ADM_ESCOLA'].value_counts().sort_index()


[1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963
 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977
 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991
 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005
 2006 2007]
Ano                             1,987
ExpecVida                          64
Populacao                  24,263,977
PIB per Capita                  8,856
PIB                   190,707,750,572
dtype: float64


count                  3,312
mean                   1,980
std                       17
min                    1,950
25%                    1,967
50%                    1,982
75%                    1,996
max                    2,007

Ano                             1,987
ExpecVida                          64
Populacao                  24,263,977
PIB per Capita                  8,856
PIB                   190,707,750,572

In [ ]:
# filtro múltiplo
mask = (data2GM['NU_ANO'] == 2015) & (data2GM['NO_MUNICIPIO_ESCOLA'] == 'Salvador') & (data2GM['NU_MEDIA_MT'] >= 600)

df.loc[mask][['NO_ESCOLA_EDUCACENSO', 'NU_MEDIA_MT']]\
.sort_values(by = 'NU_MEDIA_MT', ascending = False)\
.reset_index(drop = True)

In [ ]:
ranking_mt = data2GM_cleaned[data2GM_cleaned['ANO'].isin([i for i in range(2009, 2016)])]\
.groupby(by = ['ANO', 'SG_UF_ESCOLA'], as_index = False)['MEDIA_MT'].agg([np.mean])\
.pivot_table(index = ['SG_UF_ESCOLA'], columns = 'ANO')

ranking_mt.columns = ranking_mt.columns.droplevel(0)
ranking_mt = ranking_mt.sort_values(by = 2015)